### PR TITLE
Remove cleanupKey from registeredClosedListeners when the cache is evicted

### PR DESCRIPTION
**Pull request by luyuncheng**
### Description
we have a es cluster(31G heap, 128G Mem, 30 nodes), with many shards per node(2000 per nodes), we found in continuous read and write scenarios, heap would slowly occupied by `CompressingStoredFieldsReader` analysis shows as followings
image

__all these GC-Roots shows__:
image

### Root Cause
i think the root cause that these __threadlocal__ holds the referent, because SegmentReader#getFieldsReader calling following code, and Elasticsearch always using fixed thread_pool 
```
In `lucene/core/src/java/org/apache/lucene/index/SegmentCoreReaders.java` defined fieldsReaderLocal
  final CloseableThreadLocal<StoredFieldsReader> fieldsReaderLocal =
      new CloseableThreadLocal<StoredFieldsReader>() {
        @Override
        protected StoredFieldsReader initialValue() {
          return fieldsReaderOrig.clone();
        }
      };
```

and `IndicesRequestCache#registeredClosedListeners` always holds the `CleanupKey` unless the segment is closed. BUT even when we manually clean the request cache, the JVM heap would not release, because code
`IndicesRequestCache#cleanCache() ` only remove key from `Cache<Key, BytesReference> cache`. so it(registeredClosedListeners) would keep the referent in __threadlocal__

### Proposal
As the IndicesRequestCache#getOrCompute logic shows, when CacheEntity is evicted from cache, we can also remove the close listener and release the referent.

ISSUE: #92317

### CODE REVIEW

**Comment by benwtrent** [comment]
Jenkins test this please

**Comment by luyuncheng** [comment]
@elasticsearchmachine update branch

**Comment by luyuncheng** [comment]
@benwtrent would you pls retest it

**Comment by benwtrent** [comment]
Jenkins retest this please

**Comment by benwtrent** [comment]
@luyuncheng thank you for this change!

Wanted you to know I haven't forgotten. I am unfamiliar with this part of the codebase (as most people are). It hasn't been significantly updated in some time. 

The lack of churn and how widely used this code is making me VERY cautious, so I will have to spend some significant time digging into this change and its potential impact.

Additionally, holiday break is coming up. So, please don't be discouraged if communication on this PR goes dark for a bit.

This seems at first glance a good change and will be great to incorporate.

Thank you for your contribution and patience!

**Comment by luyuncheng** [comment]
@benwtrent  Thanks for your reply.  Happy Holiday 😊

File: server/src/main/java/org/elasticsearch/indices/IndicesRequestCache.java [hunk 1]
```diff
@@ -324,10 +324,17 @@ synchronized void cleanCache() {
         if (currentKeysToClean.isEmpty() == false || currentFullClean.isEmpty() == false) {
             for (Iterator<Key> iterator = cache.keys().iterator(); iterator.hasNext();) {
                 Key key = iterator.next();
+                CleanupKey cleanupKey = new CleanupKey(key.entity, key.readerCacheKey);
                 if (currentFullClean.contains(key.entity.getCacheIdentity())) {
+                    if (registeredClosedListeners.containsKey(cleanupKey)) {
+                        registeredClosedListeners.remove(cleanupKey);
+                    }
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/indices/IndicesRequestCache.java:331]: Here is what I am thinking.

 - `registeredClosedListeners` are only present to clear the cache if the shard is closed. 
 - When closed, `cleanupKey.entity.isOpen() == false` is true, indicating a `currentFullClean`.
 -  The only other thing that indicates a `currentFullClean` is when `readerCacheKey == null`. `readerCacheKey == null` only happens on `IndicesRequestCache#clear`

So, it make sense that the two things that indicate a `currentFullClean` (meaning, removing all cache entries with the same cacheIdentity) also removes the `onClose` handler.

I need somebody else to confirm my thinking here. 

With all this said, I think the following changes.

 - Only call `registeredClosedListeners.remove(cleanupKey);` in the `if (currentFullClean.contains(key.entity.getCacheIdentity())) {` case. Do NOT do this in the `if (currentKeysToClean.contains(cleanupKey)) {`
 - In the `currentFullClean.contains(key.entity.getCacheIdentity())` you only need to call `registeredClosedListeners.remove(cleanupKey)`. Don't bother checking `registeredClosedListeners.containsKey(cleanupKey)`, that isn't necessary.

**Review by benwtrent — CHANGES_REQUESTED** [review]
Read through this. There are two changes for sure. I still need somebody else to confirm my thinking before we move forward here.

Thanks!

**Comment by jpountz** [comment]
Good news is that this threadlocal is going away in the next Lucene major version.

Based on the issue description, I don't think that the cache is a problem here. In any case, Lucene/Elasticsearch would slowly accumulate `CompressingStoredFieldsReader` instances until there is one per segment per thread in the `search` threadpool. Changing the behaviour of the cache wouldn't help.

Elasticsearch normally doesn't allow going over 1,000 shards per node (https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/indices/ShardLimitValidator.java). Did you bump the limit or are you running on the frozen tier that configures a threshold of 3,000?

**Comment by jpountz** [comment]
In my opinion the root cause of the high memory usage was addressed by #93984, which essentially removes the threadlocal. So the remaining question is whether we think that `cleanCache` should also make sure to remove the close listeners for the cache keys that it clears.

It is indeed useless to keep listeners for shards that do not have an entry in the cache, but this change also makes the concurrency of this class harder to reason about: currently you cannot have two threads concurrently adding and removing the same entry from the `registeredClosedListeners` map, because an entry may only get added if a shard is open and removed if a shard is closed. But now it's possible that a `getOrCompute` call adds an entry while a `cleanCache` call is removing it at the same time, and I cannot figure out if it's done in a safe way.

Are there other reasons than memory usage why you are interested in this change?

**Comment by luyuncheng** [comment]
@jpountz  Thanks for your reply and consideration.

> In my opinion the root cause of the high memory usage was addressed by #93984, which essentially removes the threadlocal. 

LGTM i think this can reduce ThreadLocal memory usage

> now it's possible that a getOrCompute call adds an entry while a cleanCache call is removing it at the same time

in the `currentFullClean` , the `entity is closed` or `reader is null` , so i think it is safe to clean entry in `currentFullClean`

https://github.com/elastic/elasticsearch/blob/9a114761bf823298e21dc64353d2f2078d8fb240/server/src/main/java/org/elasticsearch/indices/IndicesRequestCache.java#L317-L320

> Are there other reasons than memory usage why you are interested in this change?

just memory reasons, i think #93984 can solve this memory issue.

**Pull request closed without merging**